src/core/indexer/mod.rs

```diff
@@ -0,0 +1,290 @@
+pub(crate) mod tables;
+
+use base64::Engine;
+
+use crate::{
+    core::indexer::tables::{BothSigCodex, CurrentSigCodex},
+    error::{Error, Result},
+};
+
+use super::util;
+
+/// Indexer is fully qualified cryptographic material primitive base class for
+/// indexed primitives. Indexed codes are a mix of indexed and variable length
+/// because code table has two char codes for compact variable length.
+pub(crate) struct Indexer {
+    /// stable (hard) part of derivation code
+    code: String,
+    /// unqualified crypto material usable for crypto operations
+    raw: Vec<u8>,
+    ///  main index offset into list or length of material
+    index: u32,
+    ///  other index offset into list or length of material
+    ondex: Option<u32>,
+}
+
+impl Default for Indexer {
+    fn default() -> Self {
+        Indexer {
+            raw: vec![],
+            code: tables::Codex::Ed25519_Sig.code().to_string(),
+            index: 0,
+            ondex: Some(u32::MAX),
+        }
+    }
+}
+
+impl Indexer {
+    pub fn new_with_code_and_raw(
+        code: &str,
+        raw: &[u8],
+        index: u32,
+        mut ondex: Option<u32>,
```

Comment:
Instead of making optional params just yet, I have usually opted for explicit functions that either have the param or don't. In the (possibly near) future, we can, if it is necessary during higher level composition, add wrappers that check for the existence of the optional and delegate to the appropriate function under the hood (think `new()` without qualifiers, and optional params just like the`__init__()` methods in Python). This gives us more flexibility in composition within CESR, and will likely allow you to clean up some if nesting in this process. Something to consider.